### ISSUE TITLE
AICE - AI Character Editor - community AIC

### ISSUE THREAD

**Issue by ByBurton** [Sh0wdown/UnofficialCrusaderPatch#280]
Hi,
I'm in charge for the AICE, and I kinda came up with an idea.
I've seen all those wishes you guys have towards all the different AI characters.
For example "improve Caliph's economy".
So I kinda want to gather all ideas in this thread, and then create a community file which we could probably attach to the patch itself then.
What do you guys think?

If you post ideas, please try to format it like this:
"AI-Lord:
-things to change 1
-things to change 2
-...
Next AI-Lord"

And try to format the things you want to change by: offence, defence, siege engines, resources, economy.

~Kimberly

**Comment by ByBurton** [comment]
@Sh0wdown Can we have a "AIC" and "Discussion" label?

**Comment by ByBurton** [comment]
I'm gonna start with my ideas:

1. Pig:
- Higher attack force increase
- More harassing macemen / more sortie macemen
- A few more crossbowmen for offence (as ranged push)
- One more quarry maybe

2. Rat:
- Decrease equipment buy amount to 1, so that the rat can always recruit units (won't have to wait for gold if it can only afford 1 bow / spear instead of 2)
- A few more units on defence
- A lot more units patrolling 
- Melee sortie units (spearmen) should be increased
- Actually restock lost defensive units
- More harassing units (just a few)

Snake:
- Very, very few arabian swordsmen as bodyguards maybe
- Change defensive units to include slingers and a few spearmen
- Add a few slingers as ranged push units
- Add a few shields to siege engines and increase the engineers accordingly

Wolf:
- Increase the defensive unit count
- Change defensive units to also include arabian archers, so he can light pitch
- Give the wolf either some crossbowmen or arabian archers as ranged push units for attacks
- Sortie unit amount (both melee and ranged) should be increased
- More siege engines (and at least one batteringram and siegetower), and increase the engineers accordingly
- (Maybe) add a few arabian swordsmen as mercenaries to his attack force
- Decrease the quarries by one, increase the iron mines by 1
- Increase patrolling unit count slightly

Wazir:
- Increase the defensive unit count
- Change defensive units to also include a few arabian swordsmen and some horse archers.
- Ranged sortie units should be horse archers, and a few more. 
- Add horse archers as ranged push units, and decrease the max amount of backup horse archers
- A few less fire throwers for attacks
- More slaves to fill up enemy moat
- More harassing horse archers
- Sell food a little later (175?)

Caliph:
- Increase defensive unit count slightly
- Change defensive units to include very few slingers and arabian swordsmen
- Ranged sortie unit (arabian archers) slight increase
- One more quarry maybe
- More harassing slaves
- Add more slaves for filling up enemy moat, and add a few to the attack force

Richard:
- Increase probability to replace defensive units to 20 or 30
- Increase defensive units very slightly
- Change harassing units to pikemen
- Add more siege engines and engineers accordingly
- More tunnelers for attacks (8 maybe?)
- Attacks should include some English longbowmen as ranged push

Saladin:
- More defensive unit count
- Increase patrolling unit count
- Increase sortie unit count
- More arabian archers as siege defence

Sultan:
- Increase defensive unit count slightly
- Change defensive units to include arabian archers and fire throwers as  well
- Add ranged push units to attack force (either very few arabian archers or a lot of slingers, preferably the later)
- One more quarry maybe

Philipp:
- No selling for pikes
- Change defensive units to include some pikemen
- Increase defensive unit count
- Change melee sortie unit to knights (2 to 4)
- Add a few pikemen to the attack force
- Add a single shield (and a single engineer) to attack force

Marshall:
- Enable buying of equipment (in amounts of 2 or 4)
- Increase sortie knights a bit
- Attack random increase a little lower

Abbot:
- Increase sortie monks
- Increase wall defence
- Defence should include some monks (monks : archers 1 : 1)
- More moat filling archers for attacks
- More raiding monks
- More patrols

Emir:
- Defensive units should include his beloved arabian swordsmen, increase total defensive unit count accordingly, so he does not lose out on fire power
- Slightly more raiding arabian swordsmen
- Slightly more sortie arabian swordsmen
- Slightly more moat filling archers for attacks

Sheriff:
- More defensive units
- Defensive units should also include macemen, not only assassins
- More patrols
- More sortie units
- A few more raiding macemen
- More slaves to fill up enemy moat
- 1 siege tower and 4 more engineers for attacks
- Crossbowmen as ranged push units

**Comment by PitchNeeded** [comment]
Raiding troops do need to be increased for most of the AIs apart from Emir and Marshall who send enough I think (although maybe as you say Emir could wait a bit until he has recruited three or so more Arabian swordsmen before he sends his raids). I do notice Emir sends more raiding Arabian swordsmen if he has a lot of gold, big lines of them. 

I would suggest Saladin would occasionally send a big group of Arabian swordsmen for raiding too, Arabian swordsmen look very regal and would fit with Saladin's status as a king. Saladin needs his counterattacking and patrolling Arabian swordsmen increased in numbers to rival Emir's I think. 

Abbot should be like Emir but with monks for raiding and counterattacking, sending big lines of monks for both. He can certainly afford it.

For Sultan yes, less woodcutters and one additional quarry would be great. He could defend his quarries with a few Arabian archers along with the current slingers and Arabian swordsmen. For Caliph, yes another quarry would be great, he definitely needs the economic boost.

You forget about Nizar and Frederick there, 

for Nizar he should build harassment siege engines and increase his raiding forces of assassins
and for Frederick larger counterattack forces and maybe build a siege tower for sieges.

It must be tested though with increasing counterattack forces, I am worried that increasing counterattack forces for AI would mean they would be slower to respond to threats on the map, as they have to train more units before they send them.

**Comment by Evrey** [comment]
> What do you guys think?

Give me an editor, give it the ability to import config files as community-created bot. I am not interested in a tool that can only select between picked community choices.

**Comment by pwnasunt** [comment]
Hello,

I welcome this idea!

I agree with Evrey's suggestion, the option to pick between custom/community profiles should be nice! If the AICE is distributed the same way as the AI Village, it will be perfect.

My suggestions:

- Lower the harassing siege engine gold threshold to 350. Currently this is set to 500 gold - bringing it lower should enable the AI to harass earlier.

**The Wolf:**

- Add pikemen or swordsmen to patrols (better defense against fast raiders)
- Add assassins to defenses and raids (complements The Wolf's profile as a strategist)
- Increase count of raiding horse archers (5 of them falls fast to defenders)
- Add fire ballistae to harassment siege engines

**The Emir:**

- Add horse archers to raids (preferrably a large group, like his Arab swordsmen)
- Add horse archers to patrols (better farm coverage, Arab swordsmen are slow to respond)
- Add fire ballistae to siege attacks (they are already made as harassing engines)
- Add a battering ram or two to siege attacks (helps Arabian swordsmen to get in)

**Emperor Frederick:**

- Set minimum count of raiding knights to 5
- Add swordsmen to patrols (better defense against fast raiders)
- Add swordsmen/knights to defensive marker types (improves wall defenses/enables 'sallying forth')

**Comment by ByBurton** [comment]
@Evrey  This tool is being worked on, by me. We will probably have an early version read this week.
You will be able to change all AIs the way you want to, but this community one will be included in the UCP.

**Comment by ByBurton** [comment]
@pwnasunt The problem with patrols is, is that you can not select specific units an AI should use just for them. They consist of the same units the defence does.
E.g. Sultan has Slingers and swordsmen as defence, what do you see around his quarries?
E.g. Wazir uses arabian archers and fire throwers as defence, what do you see as his patrols?

**Comment by Lolasik011** [comment]
@ByBurton 
Thank you for your work on the AICE tool. 

I would like to add a knight for the Wolf because I think it is in his style.

**Comment by pwnasunt** [comment]
@ByBurton How do you know that? Although it is mostly true, for example The Lionheart has pikemen with archers, yet you cannot use pikemen defense markers. Also, Saladin has a few Arabian swordsmen in his patrols, but their marker is also unusable, iirc.

**Comment by ByBurton** [comment]
@pwnasunt How I know that? I found out a lot of these variables. Richard uses pikemen and archers as defence, so that is why he has pikemen as patrols.
Same for Saladin.
Here, the part of Richard's defence I am talking about:
``DefUnit1	= EuropArcher
		DefUnit2	= EuropArcher
		DefUnit3	= Pikeman``

**Comment by pwnasunt** [comment]
@ByBurton Interesting. It seems that sortie units also qualify as defenders, right? :D I thought defenders == static defenders only.

**Comment by ByBurton** [comment]
@pwnasunt Nope. Sortie units are completely standalone. They have nothing to do with defence.
Defenders can be split in wall defenders and patrols, but use the same unit types.

**Comment by Truetobi91** [comment]
I'm somewhat irritated by many of your suggestions as I thought like half of them are actually the standard. The other half are changes I'd not necessarily support and we have to differentiate between some kinds of players modding their game:

(a) likes to improve their AI as good as possible
(b) likes to moderately improve the AI while maintaining the vanilla strength scale of AI lords
optional (c) likes to slightly improve AI behaviour without changing too much stuff - improved vanilla to say so.

**Comment by ByBurton** [comment]
@Truetobi91 I have made a aic file which maintains the vanilla balance as close as possible.
Although I improved the pig a slight little bit more than others, as he is supposed to be stronger than the Snake, and he is underpowered. I also slightly made the wolf even slower with attacking. But with more force. (In the original campaign he also only acted in the end)

This one here should be different, it should be made out of community ideas, and not focused on any of your ideas. It should be a mix out of all the different ideas.

**Comment by PitchNeeded** [comment]
> **The Emir:**
> 
> * Add horse archers to raids (preferrably a large group, like his Arab swordsmen)
> * Add horse archers to patrols (better farm coverage, Arab swordsmen are slow to respond)

Hmm, I don't think this would work well, he would take ages to raise a siege force, and horse archers to patrols would make him too similar to Wazir. At the minute, I notice that Emir first sends Arabian archers out to archers to respond to attacks on his farms and if these aren't effective in dealing with the threat then then he sends the Arabian swordsmen. I think those tactics are OK as they are.

But I do notice already Emir sometimes sends out a few horse archers to deal with enemy troops attacking his walls. If he could send out some more and not use them for melee combat as he does now but keep them in a safe place just outside his castle firing at the melee troops hacking at his walls, this would be great! Wazir and Saladin could also use the same tactics.

> **The Emir:**
> 
> * Add fire ballistae to siege attacks (they are already made as harassing engines)
> * Add a battering ram or two to siege attacks (helps Arabian swordsmen to get in)

Good ideas for Emir. A ram would be good for Emir for sieges, I do see him almost as the Arabian equivalent of Richard with similar castle defences and reliance on heavy infantry for attack and defence, so it would good if he could be like Richard and use a ram or two.

> Emperor Frederick:
> 
> Add swordsmen to patrols (better defense against fast raiders)
> Add swordsmen/knights to defensive marker types (improves wall defenses/enables 'sallying forth')

Frederick already sends 4 swordsmen to counter attack sometimes. He needs to send more than just 4, as this is not enough I think.

**Comment by PitchNeeded** [comment]
Now the brilliant developers seem to have deciphered the coding needed to edit AI behaviour, would it be possible eventually to create a new AI character? Considering if you edit an AI's behaviour enough using the AICE it would become a new AI essentially, all that would be required is to create a space in the game to have this new AI instead of replacing another one. Yes there would be no speech files or binks initially but I am sure eventually these could be created by someone.

**Comment by ByBurton** [comment]
"Creating space" isn't that easy probably. But yeah, would be awesome! I still fancy my idea of Robin Hood, using weak units like the rat, but the other way around. Spamming archers, building those watchtowers The Caliph fancies so much, and having kind of a mix of a weird castle in the midst of a nizar island :D .
🤷‍♀️

**Comment by PitchNeeded** [comment]
> "Creating space" isn't that easy probably. But yeah, would be awesome! I still fancy my idea of Robin Hood, using weak units like the rat, but the other way around. Spamming archers, building those watchtowers The Caliph fancies so much, and having kind of a mix of a weird castle in the midst of a nizar island :D .
> 🤷‍♀️

Sounds great, but rather than spearmen, I would love a Robin Hood character to use tunnelers for melee combat in siege along with the archers, it would be unique, and also tunnelers are faster than spearmen and have a decent attack.

Still would love to see Lord Woolsack from Stronghold 1 with a mix of pikemen, monks and archers. Lord Woolsack was unfairly left out of the game I think when Sir Longarm is there!

**Comment by Lolasik011** [comment]
> Still would love to see Lord Woolsack from Stronghold 1 with a mix of pikemen, monks and archers. Lord Woolsack was unfairly left out of the game I think when Sir Longarm is there!


I would also like to see Lord Woolsack. All the more talk animation and sounds are in Stronghold 1.

**Comment by PitchNeeded** [comment]
> I would also like to see Lord Woolsack. All the more talk animation and sounds are in Stronghold 1.

We would just have to make it so you have to have Stronghold 1 to have that character and use the audio files, but yes that wouldn't be a problem. He would be the pikemen specialist to combat his old rival Sir Longarm's knights! And he would have monks because he looks like a monk.

But the game actually having a female character would be excellent too, and for this Eleanor of Aquitaine would be the best candidate! A queen consort who took part in the Second Crusade!

**Comment by Panbutt** [comment]
This is amazing to hear! While I'm bristling with ideas, I'll keep them to a minimum so I don't come off as overwhelming aha. Most of my suggestions here are more quality of life adjustments:

- Add a gold threshold for weapon purchasing of 1,000 gold. Example: an AI will not buy in weapons unless they have more than 1,000 gold. This'll help prevent AI from starving themselves like Marshal, Rat, Snake and Lionheart. It leaves them with some spending/emergency money. This will also prevent AIs from wasting all of their starting gold on weapons at the start of a game.
- Remove the missile trooper limit for attack forces, this will help with countering castle defenses more reliably.

**Pig**
- Remove his one pitch rig and replace it with an extra iron mine (he sells pitch immediately so 1 pitch rig is useless)
- Let the poor man sell his cheese surplus.

**Nizar**
- Add Arabian archers to his attack forces. (His assassins die way too quickly without missile cover)
- Add harassing fire ballistae (3 max?)

**Wazir**
- Increase wheat farm count from 8 max to 12 max.

**Phillip**
- Let him build at least 1 quarry.

**Comment by ByBurton** [comment]
@Panbutt I like your ideas for Philipp and Wazir.

**Comment by PitchNeeded** [comment]
**Marshall, Emir, Saladin, Lionheart, Pig, Frederick**

- Two more iron mines

**Caliph**

- Another iron mine and another quarry

**Rat**

Three more apple farms

**Sultan**

One more quarry and 4 less woodcutters

**Phillip, Nizar and Snake**

One more quarry

> Wazir
> 
> Increase wheat farm count from 8 max to 12 max.

One more iron mine and two more wheat farms would be better. I think 10 is enough with negative fear factor and the additional iron mine would help him to do better on a map with little oasis area.

**Comment by PitchNeeded** [comment]
I am worried though about the possible impact of overpopulation for ale-using AI (going from +8 to +6 ale bonus) if we add more economic buildings to AIs but the bonuses from having more production should outweigh the negatives from this (I think)

**Comment by ByBurton** [comment]
@PitchNeeded That would strengthen the AI way too much, consider maybe something less strong.

**Comment by PitchNeeded** [comment]
> @PitchNeeded That would strengthen the AI way too much, consider maybe something less strong.

OK maybe 1 additional iron mine for iron mine AIs instead of two more....the exception being Caliph who needs a bit more of an economic boost (1 more iron mine and 1 more quarry) because he is often so slow raising new siege forces because of his expensive troops. I know you said before he is only supposed to be a medium difficulty AI but I think he would often struggle against Snake at the minute because of his poor economy.

**Comment by Monsterfisch** [comment]
am i right to assume if the ai is assigned units it doesnt have workshops for will automatically buy in the required equipment?

**Comment by ByBurton** [comment]
@Monsterfisch yes. Regarding workshops though: I have no idea how to change what a workshop produces. For example I have no idea how to tell Philipp to produce pikes instead of spears.

**Comment by Monsterfisch** [comment]
mhm i see that sure is a bit of a problem but i am sure that after enough digging a value for that will surface eventually. As long as the ai buys in the required equipment that a good start to get some more balanced ais going particularly with the abbot and his gold storage. that being said really looking forward to trying out setups myself with the editor you are working on. particulary with the option to boost the aiv economy to make up for the increased need of gold in case they need to buy in additional armor and weapons to be able to recruit those special "new" units

**Comment by PitchNeeded** [comment]
> mhm i see that sure is a bit of a problem but i am sure that after enough digging a value for that will surface eventually. As long as the ai buys in the required equipment that a good start to get some more balanced ais going particularly with the abbot and his gold storage. that being said really looking forward to trying out setups myself with the editor you are working on. particulary with the option to boost the aiv economy to make up for the increased need of gold in case they need to buy in additional armor and weapons to be able to recruit those special "new" units



> "Creating space" isn't that easy probably. But yeah, would be awesome! I still fancy my idea of Robin Hood, using weak units like the rat, but the other way around. Spamming archers, building those watchtowers The Caliph fancies so much, and having kind of a mix of a weird castle in the midst of a nizar island :D .
> 🤷‍♀️

I was thinking that until a space is created for a new AI in the game, we should have an optional feature as part of the AICE to temporarily swap any one of the other AIs with any player made AI for a game so we can just test it out and experiment.

I might start creating AIV files for my Lord Woolsack proposal for when AICE is out. I think a castle with lots of small perimeter turret towers and one or two big square towers could be Woolsack's style. No vanilla AIVs use lots of perimeter turrets, and they look quite nice, and would make him unique. No pitch (not his character), but he would use some wooden traps. Quite a few poleturners for his pikemen, a cathedral, inns and statues.

**Comment by Monsterfisch** [comment]
100% agree i will certainly try my hand at prototyping some new characters (economy and unit wise) that would be interesting to see in game in hope that it will be possible to get them in there additionally.

**Comment by Lolasik011** [comment]
I can not wait AICE ...

**Comment by Lolasik011** [comment]
Really looking forward to AICE ....

**Comment by PitchNeeded** [comment]
> Really looking forward to AICE ....

Indeed it's incredible that editing AI tactics, economic buildings and unit types is possible.

**Comment by Sh0wdown** [comment]
IF you already want to get some insight, here is such a file.
[vanilla.zip](https://github.com/Sh0wdown/UnofficialCrusaderPatch/files/3010836/vanilla.zip)

The AICE will provide a GUI so you don't have to edit text files, though you can still do that if you want.
You will be able to install your own edited .aic files with the UCP. I would like to include a few .aics from the community optionally in the launcher. Maybe one can even combine AIC & AIV edits.

I'm also looking into the addition of ne AIs, but that will take more time.

**Comment by PitchNeeded** [comment]
> 100% agree i will certainly try my hand at prototyping some new characters (economy and unit wise) that would be interesting to see in game in hope that it will be possible to get them in there additionally.

Also if we are to experiment to make new AIs with the AICE, can it be made please, if possible, so we can edit AI sound files and binks with the AICE, using any Stronghold Crusader sound files as quotes (even peasant quotes!) and using any Stronghold Crusader binks as videos like the halberdier one (from the campaign) and the one with the jester (used for the sandbox/invasion scenario travelling fair event)?

**Comment by Monsterfisch** [comment]
> I'm also looking into the addition of ne AIs, but that will take more time.

pretty sure this implies the additional soundfiles etc =)

**Comment by Truetobi91** [comment]
>
> The [AICE](https://github.com/ByBurton/AI_Character_Editor) will provide a GUI so you don't have to edit text files, though you can still do that if you want.
> You will be able to install your own edited .aic files with the UCP. I would like to include a few .aics from the community optionally in the launcher. Maybe one can even combine AIC & AIV edits.
> 
> I'm also looking into the addition of ne AIs, but that will take more time.

A related feature that I'd like to wish is a textfile that will be created after installing and includes all the changes that you made with the patch which makes it easier to track down if you forgot something or did something wrong while installing or can report a bug more reliably. 

Optionally there could also be a file that let's you quickly install your last preset of changes e.g. one for active gameplay and one for AI battles when you feel the need for different settings in both scenarios.

**Comment by ByBurton** [comment]
@Truetobi91 You could open up an issue for that idea.

**Comment by PitchNeeded** [comment]
**Phillip** 

An option in the patch once AICE is released to let him use a mix of knights, spearmen, archers and catapults in a siege.

**Comment by ByBurton** [comment]
@PitchNeeded It would look stupid though, seeing like 6 knights attacking.... And stables are expensive.

**Comment by Monsterfisch** [comment]
The only thing those knights would look stupid doing is camping in front of the keep killing peasants over and over again to cripple the enemies economy even further to buy more time to recruit the finishing force ;)

**Comment by ByBurton** [comment]
@Monsterfisch Knights aren't that good ... I won't create an aic with philipp using knights as attack force, it would be way to expensive for him, and ineffective as well.

**Comment by Monsterfisch** [comment]
i am not trying to convince anyone to put knights in the aic after all that can be tried out by people themselves later if they are curious all i am saying is that knight stats wise sure look better than spearman and if Phillips economy can be upped a little bit the price wouldn't be a big issue anymore. and as it has been pointed out before Phillip sure likes to boast with his knightly strikeforce it would fit his character

**Comment by ByBurton** [comment]
Yes, I know, I know. And as you said, you can put those knights in yourself. But even if he boasts with them, having a max of, say, 8 knights "charge" into battle will always more look like a suicide squad. They won't be a threat to anyone probably. They will act like any regular meele units in an offensive force, attack a wall (or towers / gates with the UCP). And that is so not the "ride them down" Philipp meant.

**Comment by Monsterfisch** [comment]
i hear you but that is exactly why they should be added to the attack force as the knight will at least get to the walls and deal damage plus buy time for the spearman to get closer because they have much better stats, anyways no need to debate this any further after all its a minuscule thing to get worked up about ;)

**Comment by Rakso69** [comment]
This stuff is a really great thing! I'll edit my own Wolf once it get's published. I hope it to have under access every single AI aspect, not limited by any case, by which I also mean min/max catapults/trebuchets/fire balista etc. for sieges, min/max woodcutters and alike. Basically every single one of these if that's possible to edit.
By the way I'd like to ask: I've noticed some users use some different AIV Editors over here(based on pics they post), looks it has a desert and has a higher resolution, the one on which I'm working uses grass and low resolutions. Could anyone post a link to a "best" or "better" one if has one for download? Maybe it allows for some additional guarding troops to be deployed, also more cata/trebuchets too? Feeling like I would might still be using old tools, while new ones are avaible for use already, lmao! Cheers!
